src/coprocessor/dag/executor/index_scan.rs

```diff
@@ -53,45 +56,54 @@ impl<'a> IndexScanExecutor<'a> {
             pk_col = Some(cols.pop().unwrap());
         }
         let col_ids = cols.iter().map(|c| c.get_column_id()).collect();
-        let scanner = Scanner::new(store, desc, false, statistics);
 
         COPR_EXECUTOR_COUNT.with_label_values(&["idxscan"]).inc();
         IndexScanExecutor {
+            store: store,
+            statistics: Statistics::default(),
             desc: desc,
             col_ids: col_ids,
-            scanner: scanner,
-            key_ranges: key_ranges,
-            cursor: Default::default(),
             pk_col: pk_col,
+            key_ranges: key_ranges,
+            cursor: 0,
+            scanner: None,
+            scan_key_only: false,
         }
     }
 
     pub fn new_with_cols_len(
         cols: i64,
         key_ranges: Vec<KeyRange>,
-        store: SnapshotStore<'a>,
-        statistics: &'a mut Statistics,
-    ) -> IndexScanExecutor<'a> {
+        store: SnapshotStore,
+    ) -> IndexScanExecutor {
         let col_ids: Vec<i64> = (0..cols).collect();
         COPR_EXECUTOR_COUNT.with_label_values(&["idxscan"]).inc();
-        let scanner = Scanner::new(store, false, false, statistics);
         IndexScanExecutor {
+            store: store,
+            statistics: Statistics::default(),
             desc: false,
             col_ids: col_ids,
-            scanner: scanner,
-            key_ranges: key_ranges,
-            cursor: Default::default(),
             pk_col: None,
+            key_ranges: key_ranges,
+            cursor: 0,
+            scanner: None,
+            scan_key_only: false,
         }
     }
 
     pub fn get_row_from_range(&mut self) -> Result<Option<Row>> {
-        let range = &self.key_ranges[self.cursor];
+        let range = mem::replace(&mut self.key_ranges[self.cursor], KeyRange::default());
         if range.get_start() > range.get_end() {
             return Ok(None);
         }
-        let kv = self.scanner.next_row(range)?;
-        let (key, value) = match kv {
+
+        if self.scanner.is_none() {
+            let scanner = Scanner::new(&self.store, self.desc, self.scan_key_only, range)?;
```

Comment:
Create a scanner is much more expensive than set seek key. It will become a problem if there is many ranges.